contracts/contracts/GrantRegistry.sol

```diff
@@ -0,0 +1,135 @@
+pragma solidity ^0.8.5;
+
+contract GrantRegistry {
+  // Grant ID - The hash of the grants meetadata
+  bytes32 public grantRegistryId;
+  // Metadata Pointer
+  string public metaPtr;
+
+  // Owner
+  address owner;
+
+  modifier onlyOwner() {
+    require(msg.sender == owner);
+    _;
+  }
+
+  // TODO: Only allow the owner (which is the curator) address to configure the contract
+  // while its in config mode, then switch to active. Once active, the owner can only
+  // close the registry (have a conversation around this approach)
+  constructor(
+    address _owner,
+    bytes32 _grantRegistryId,
+    string memory _metaPtr
+  ) {
+    owner = _owner;
+    //state = State.CONFIG;
+    grantRegistryId = _grantRegistryId;
+    metaPtr = _metaPtr;
+  }
+
+  // State
+  enum State {
+    CONFIG,
+    ACTIVE,
+    CLOSED
+  }
+  //State public state;
+
+  // Grant Pointer Object
+  // Mapping from GrantID to set of Owner Address, Payout Address, Metadata Pointer, Replaces Grant (optional)
+  // TODO: Implement safemath for avoiding overflows with ints
+  uint256 public grantCount = 0;
+  mapping(bytes32 => Grant) public grants;
+  bytes32[] public grantIds;
+
+  struct Grant {
+    bytes32 id;
+    address owner;
+    address payout;
+    string metaPtr;
+    State state;
+    bytes32 replaces;
+  }
+
+  function activate(bytes32 _id) external onlyOwner {
+    grants[_id].state = State.ACTIVE;
+  }
+
+  function closed(bytes32 _id) public onlyOwner {
+    grants[_id].state = State.CLOSED;
+  }
+
+  function isConfig(bytes32 _id) public view returns (bool) {
+    return grants[_id].state == State.CONFIG;
+  }
+
+  function isActive(bytes32 _id) public view returns (bool) {
+    return grants[_id].state == State.ACTIVE;
+  }
+
+  function isClosed(bytes32 _id) public view returns (bool) {
+    return grants[_id].state == State.CLOSED;
+  }
+
+  function getGrantCount() public view returns (uint256) {
+    return grantCount;
+  }
+
+  function getAllGrantIds() public view returns (bytes32[] memory) {
+    bytes32[] memory ret = new bytes32[](grantCount);
+    for (uint256 i = 0; i < grantCount; i++) {
+      ret[i] = grantIds[i];
+    }
+    return grantIds;
+  }
+
+  function getAllGrants() public view returns (Grant[] memory) {
+    Grant[] memory ret = new Grant[](grantCount);
+    bytes32 grantId;
+    for (uint256 i = 0; i < grantCount; i++) {
+      grantId = grantIds[i];
+      ret[i] = grants[grantId];
+    }
+    return ret;
+  }
```

Comment:
Should be able to save some gas by removing an MSTORE—even though it's an off-chain helper, you can still run out of gas because nodes have a gas limit on `eth_call`s. For example (using a `returnData` to be more explicit than `ret`):

```solidity
  Grant[] memory returnData = new Grant[](grantCount);
  for (uint256 i = 0; i < grantCount; i++) {
    returnData[i] = grants[grantIds[i]];
  }
  return returnData;
```

Also, given that this can run out of gas, there should also be a version of this method that lets you enter a start and end index, because eventually this call will fall when `grantCount` is large enough. For example

```solidity
function getAllGrants() external view returns (Grant[] memory) {
    return getSomeGrants(0, grantsCount);
}

function getSomeGrants(uint256 startIndex, uint256 endIndex) public view returns (Grant[] memory) {
    Grant[] memory returnData = new Grant[](grantCount);
    bytes32 grantId;
    for (uint256 i = startIndex; i < endIndex; i++) {
      grantId = grantIds[i];
      returnData[i] = grants[grantId];
    }
    return returnData;
}
```